packages/common/src/search-attributes.ts

```diff
@@ -0,0 +1,300 @@
+    for (const [key, attr] of Object.entries(this.searchAttributes)) {
+      const attrKey = { name: key, type: attr.type };
+      // Sanity check, should always be legal.
+      if (isValidValueForType(attrKey.type, attr.value)) {
+        res.push({ key: attrKey, value: attr.value } as SearchAttributePair);
+      }
+    }
+    return res;
+  }
+
+  static getKeyFromUntyped(
+    key: string,
+    value: SearchAttributeValueOrReadonly
+  ): SearchAttributeKey<SearchAttributeType> | undefined {
+    if (value == null) {
+      return;
+    }
+
+    // Unpack single-element arrays.
+    const val = value.length === 1 ? value[0] : value;
+    switch (typeof val) {
+      case 'string':
+        // Check if val is an ISO string, if so, return a DATETIME key.
+        if (!isNaN(Date.parse(val)) && Date.parse(val) === new Date(val).getTime()) {
+          return { name: key, type: SearchAttributeType.DATETIME };
+        }
+        return { name: key, type: SearchAttributeType.TEXT };
+      case 'number':
+        return {
+          name: key,
+          type: Number.isInteger(val) ? SearchAttributeType.INT : SearchAttributeType.DOUBLE,
+        };
+      case 'boolean':
+        return { name: key, type: SearchAttributeType.BOOL };
+      case 'object':
+        if (val instanceof Date) {
+          return { name: key, type: SearchAttributeType.DATETIME };
+        }
+        if (Array.isArray(val) && val.every((item) => typeof item === 'string')) {
+          return { name: key, type: SearchAttributeType.KEYWORD_LIST };
+        }
+        return;
+      default:
+        return;
+    }
+  }
+
+  static toMetadataType(type: SearchAttributeType): string {
+    switch (type) {
+      case SearchAttributeType.TEXT:
+        return 'Text';
+      case SearchAttributeType.KEYWORD:
+        return 'Keyword';
+      case SearchAttributeType.INT:
+        return 'Int';
+      case SearchAttributeType.DOUBLE:
+        return 'Double';
+      case SearchAttributeType.BOOL:
+        return 'Bool';
+      case SearchAttributeType.DATETIME:
+        return 'Datetime';
+      case SearchAttributeType.KEYWORD_LIST:
+        return 'KeywordList';
+      default:
+        throw new Error(`Unknown search attribute type: ${type}`);
+    }
+  }
+
+  static toSearchAttributeType(type: string): SearchAttributeType | undefined {
+    switch (type) {
+      case 'Text':
+        return SearchAttributeType.TEXT;
+      case 'Keyword':
+        return SearchAttributeType.KEYWORD;
+      case 'Int':
+        return SearchAttributeType.INT;
+      case 'Double':
+        return SearchAttributeType.DOUBLE;
+      case 'Bool':
+        return SearchAttributeType.BOOL;
+      case 'Datetime':
+        return SearchAttributeType.DATETIME;
+      case 'KeywordList':
+        return SearchAttributeType.KEYWORD_LIST;
+      default:
+        return;
+    }
+  }
+}
+
+export function defineSearchKey<T extends SearchAttributeType>(name: string, type: T): SearchAttributeKey<T> {
```

Comment:
Sorry, I wasn't clear… In my mind, this would have been the function named `defineSearchAttribute()`, but I see that "search attribute" could equivalently mean the key or the pair, so not not ideal.

Still, I'd go for `defineSearchAttributeKey()`. "Search Attribute" is a concept, something that someone is likely to search for. Reducing it to a shorter form might make it harder for users to find docs.

Reply:
Renamed to `defineSearchAttributeKey`